### DISCUSSION TITLE
Serving (watching) an application with dependent libs without building deps (linking to src files instead of build artifacts)

### DISCUSSION THREAD

**Discussion by unleashit** [nrwl/nx#17002] [General]
I have an integrated style typescript monorepo with a Node application and a few Node library dependencies. I'd like to be able to run a watch process on the app + deps. Doesn't matter much if behind the scenes it's ts-node or an alternative, but I'd prefer not to have to run the build on the libraries and instead hotlink to the source files to free up CPU cycles on an older laptop. I have this setup on another project that uses pnpm workspaces and it works fine.

Here are the relevant parts of the App's project.json (created by generator):

```json
{
  "name": "demo-server",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "apps/demo-server/src",
  "projectType": "application",
  "targets": {
    "build": {
      "executor": "@nx/esbuild:esbuild",
      "outputs": ["{options.outputPath}"],
      "defaultConfiguration": "production",
      "options": {
        "platform": "node",
        "outputPath": "dist/apps/demo-server",
        "format": ["cjs", "esm"],
        "bundle": false,
        "main": "apps/demo-server/src/main.ts",
        "tsConfig": "apps/demo-server/tsconfig.app.json",
        "assets": ["apps/demo-server/src/assets"],
        "generatePackageJson": true,
        "esbuildOptions": {
          "sourcemap": true,
          "outExtension": {
            ".js": ".js"
          }
        }
      },
      "configurations": {
        "development": {},
        "production": {
          "esbuildOptions": {
            "sourcemap": false,
            "outExtension": {
              ".js": ".js"
            }
          }
        }
      }
    },
    "serve": {
      "executor": "@nx/js:node",
      "defaultConfiguration": "development",
      "options": {
        "buildTarget": "demo-server:build"
      },
      "configurations": {
        "development": {
          "buildTarget": "demo-server:build:development"
        },
        "production": {
          "buildTarget": "demo-server:build:production"
        }
      }
    }
}
```
and a library (also created by generator) : 

```json
{
  "name": "server",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/server/src",
  "projectType": "library",
  "targets": {
    "build": {
      "executor": "@nx/js:tsc",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/libs/server",
        "tsConfig": "libs/server/tsconfig.lib.json",
        "packageJson": "libs/server/package.json",
        "main": "libs/server/src/index.ts",
        "assets": ["libs/server/*.md"]
      }
    }
}
```
main nx.json: 

```json
{
  "$schema": "./node_modules/nx/schemas/nx-schema.json",
  "npmScope": "react-help-desk",
  "tasksRunnerOptions": {
    "default": {
      "runner": "nx/tasks-runners/default",
      "options": {
        "cacheableOperations": [
          "build",
          "lint",
          "test",
          "e2e",
          "build-storybook"
        ]
      }
    }
  },
  "targetDefaults": {
    "build": {
      "dependsOn": ["^build"],
      "inputs": ["production", "^production"]
    },
    "test": {
      "inputs": ["default", "^production", "{workspaceRoot}/jest.preset.js"]
    },
    "lint": {
      "inputs": [
        "default",
        "{workspaceRoot}/.eslintrc.json",
        "{workspaceRoot}/.eslintignore"
      ]
    },
    "e2e": {
      "inputs": ["default", "^production"]
    },
    "build-storybook": {
      "inputs": ["default", "^production", "!{projectRoot}/.storybook/**/*"]
    }
  },
  "namedInputs": {
    "default": ["{projectRoot}/**/*", "sharedGlobals"],
    "production": [
      "default",
      "!{projectRoot}/**/?(*.)+(spec|test).[jt]s?(x)?(.snap)",
      "!{projectRoot}/tsconfig.spec.json",
      "!{projectRoot}/jest.config.[jt]s",
      "!{projectRoot}/.eslintrc.json",
      "!{projectRoot}/**/*.stories.@(js|jsx|ts|tsx|mdx)",
      "!{projectRoot}/.storybook/**/*"
    ],
    "sharedGlobals": ["{workspaceRoot}/babel.config.json"]
  },
  "generators": {
    "@nx/react": {
      "application": {
        "style": "scss",
        "linter": "eslint",
        "bundler": "vite",
        "babel": true
      },
      "library": {
        "style": "scss",
        "linter": "eslint",
        "unitTestRunner": "vitest"
      },
      "component": {
        "style": "scss"
      }
    }
  }
}

```

All library's are added as paths to the base tsconfig.json, yet if I delete the dist folder of the libraries I see that the app is in fact linking to the dist folders, and not the src: `Error: Cannot find module '/home/jg/www/react-help-desk/dist/libs/server'`

What I've done:

1. added main, module and exports fields into the library `package.json` to point to the source files instead of dist. Not sure if NX overwrites any of these when it merges `project.json`.
2. I deleted the "build" task in a library and everything worked as expected! But it's a buildable library and I need that task when I want to explicitly build.

I'm not completely clear yet on how the @nx/js:node executor works, but I see that it calls the build tasks of both itself and deps. The watch processes are added to the libraries seemingly as a part of NX's dependency graph, but I'm not sure how I would go about keeping everything as-is in regards to building, but not when I'm in dev mode I'd prefer to only rebuild the main app as I go. Any ideas? Thanks!

**Comment by unleashit** [comment]
I found by adding a `null` outputPath to the build tasks for dev mode (while keeping the default/prod going to dist), it all "works" but with intermittent errors:

```json
    "build": {
      "executor": "@nx/js:tsc",
       // ...
      "configurations": {
        "development": {
          "outputPath": null
        }
      }
    }
```
Now with `nx app:serve`, app is rebuilt with changes and libraries are being watched with (I assume) tsc --watch --noEmit.  This is what I was looking for, but something isn't right as it chokes every few saves and I have to restart the tasks.

2 cents tip for the maintainers: I spent a fair amount of time digging into the docs and NX source code, the latter which I found to be pretty deep in abstractions and at times coupled with Angular or Angular style (ex: rxjs). Not saying that's bad, but for me it was time consuming to grok. The docs aren't a big help and are pretty shallow when it comes to customizing the configs that get generated. AFAICT there's not even a way to debug the actual final commands that get called. It just says the name of the NX command and the output of the script (not the final command like `tsc file --args` or `node file --args`, `next start`, etc.)

Even though I have a fair amount of experience with both monorepos and build tools, I think the idea of being able to quickly generate tooling boilerplate is awesome. But only if it's possible to pretty quickly get a handle on customizing it. If I'm being honest, I'm not ready to dedicate days and days earning another proprietary Javascript build tool system just to get it to bend to my will. I think it's great that NX can be more easily used in a package based way and will probably revert to that. But I'll be looking forward to see if some energy gets put into better docs and making the integrated style more flexible.

**Comment by unleashit** [comment]
Interesting. Not sure if this post had anything to do with it but new schema enforcement was just added in a recent update. I'm now getting `Property 'outputPath' does not match the schema`, so the hack no longer works!

A null outputPath was never the answer to what I was trying to do anyway, but it confuses me when trying to understand how NX executors like @nx/js:tsc call commands since a null (or missing) path is valid for the tsc command. I get that certain configurations won't make sense in the context of a given NX setup, but this would imply tsc --noEmit isn't supported. Not sure why not, nor am I really sure from the docs how to choose and combine executors to get the desired end result of a lot of things.. 

Just trying to reason about it but, guess everyone is busy. What I was really striving was simply serving an app so the app rebuilds both on changes to itself and also any dependent packages (libs) but without having to also rebuild the libs. If someone could confirm this can or cannot be achieved through executors I'd still appreciate it, but I've already switched gears.

**Comment by clancywist** [comment]
Hello @unleashit 😃 
Did you find a workaround ? I would like to do the same things

**Reply by unleashit** [reply]
Hey sorry for the late reply. Not sure if it will work in your case, but I did actually end up finding a pretty good solution... but it only works for publishable libs.

Basically the idea is to target your source files directly in package.json with the usual fields like `main` and `exports`, but then override at publish time with built artifacts using the [publishConfig](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#publishconfig) property.  During development of the app, since imported libs will point straight to the source files (including types), there's no need to build. ~~Editing them should also make the app process live reload if set up as dependencies in nx.json~~

Here's how I have my package.jsons:

```javascript
  "main": "src/index.ts",
  "types": "src/index.d.ts",
  "exports": {
    ".": "./src/index.ts",
    "./dist/*.module.css": "./src/scss/*.module.scss"
  },
  "publishConfig": {
    "main": "dist/index.js",
    "module": "dist/index.mjs",
    "types": "dist/index.d.ts",
    "exports": {
      ".": {
        "import": "./dist/index.mjs",
        "require": "./dist/index.js"
      },
      "./dist/*.css": "./dist/*.css"
    },
    "access": "public"
  },
  "sideEffects": [
    "*.css",
    "*.scss"
  ],
``` 

Until you run an `npm tar` or `npm publish`, publishConfig` does nothing. But when publishing, a new package.json gets written in the tarball that will override the top level props, and then strip out the pubishConfig. There are a few caveats:

1. Only works with publishable libs
2. At least as of a few months ago, when using vanilla npm, publishConfig ignored the `main` field. But no issues with pnpm (and at least so I read, yarn).
3. I'm now working with an "package based" monorepo (nx not managing package.json). I'm sure it's possible, but I'm not exactly sure how you'd rig this in the integrated style.

EDIT: it's been a few months since I set this up, so I forgot you also have to make sure you are not excluding the paths to the library source files in whatever bundler (Vite, Webpack, etc.) you're using in the app(s) in dev mode. Then in production mode they should be excluded to make sure they don't get included in the app's build. So it's somewhat complicated, but it works beautifully and fast for the particular project.

**Reply by Yuripetusko** [reply]
@unleashit this sounds like a great setup and similar to what I am looking for for my monorepo with publishable libraries. Would you mind sharing example of vite config for some of your publishable libraries with this pattern? (if you are using vite and modified vite.config). thanks

**Reply by unleashit** [reply]
This is actually kind of an older project that uses Webpack, not Vite. Here's the [repo](https://github.com/unleashit/npm-library/tree/master) and the app in question's [wepack.config](https://github.com/unleashit/npm-library/blob/master/demos/frontend/webpack.config.js).

The config has since changed. Unfortunately, after examining the bundle with webpack bundle analyzer I saw I was still including the external lib code in the bundle.  Then once I did get it set up correctly to include conditionally in dev (and not in prod), it broke the builds since the package.jsons in the libs are pointing to the src files which includes unbuilt Typescript.

The only way I can think of to make this work properly is to write a script the transforms the package.json files before you build. publishConfig manages this at publish time, but otherwise your main/module/exports/type fields will still be pointing at the src files when you build. Could be worth doing, because the DX really is good. Before, with all the watch processes the whole thing was terribly sluggish.

But without a solution like that, I'm afraid I can't recommend this after all. In my case, I'm leaving for now... but only because the libs are very small and it's basically just a home made Storybook for previewing a component library.

EDIT:

Just realized (I don't work on this often) the problem isn't exactly the code getting bundled from the libs. If you're bundling the app, that actually needs to happen. It's more about what is processing it. If you directly include the source from external packages, it will be built/processed only by the app's build tools. Maybe that's actually fine, but if you do particular transformations when you build the libs, they will never get done. So still hacky.

**Reply by Yuripetusko** [reply]
After trying this for a day, I actually decided to switch to pnpm workspaces for a monorepo with publishable libraries. I love nx, but it's too much work to make it good with publishable libraries. pnpm workspaces been a breeze

**Reply by unleashit** [reply]
If I can ask, how do you have it set up? Basically, that's what I'm using here. Nx is just for caching. While I'm using NX commands to run the dev and build scripts, underneath it's just pnpm workspaces. Aside from the publishConfig config trick (or scripting to change the package.json fields) I haven't figured out way around running watch processes for each dependency while developing.

**Reply by Yuripetusko** [reply]
Yep for developing mode I still need to add watch script that will watch all libraries recursivly and rebuild them for hot-reload to work when working on the app that consumes them or when working on one of the libraries that consumes others. I am planning to work on this tomorrow, but I think just a simple vite build with watch flag will do and then prepend this script before `dev` script basically. Although it might slow down rebuilding times when developing, i'll report back when I try.

You can follow the progress here if you want: https://github.com/rmrk-team/rmrk-js and here's a TODO checklist I have for tomorrow https://github.com/rmrk-team/rmrk-js/issues/1

**Reply by Yuripetusko** [reply]
I've seen another set up in the wild where during development the script would go through each package.json and replace all main/files/export fields to point to `src` instead of `dist`, but I thought this is an overkill, although this would not impact local dev as much as `watch` would

**Reply by Yuripetusko** [reply]
This is the project I've seen in the wild that does the trick with modifying all package.json paths during development
https://github.com/wevm/wagmi/blob/main/.scripts/preconstruct.ts

**Reply by unleashit** [reply]
Ah ok, that's pretty much what I was trying to avoid. After working on a bunch of monorepos, I can't tell you how much nicer the publishConfig setup has made this one. 

Yes, I think the next time I set one up, I might go with scripting the package.jsons either like they did, or maybe using something like jinja2 templates.
